File: app/src/main/java/com/fruitblast/game/GameView.java
package com.fruitblast.game;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.RectF;
import android.os.Handler;
import android.os.Looper;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;

public class GameView extends View {

    // ── Constructors (all 3 required for XML inflation) ───────────────────────

    public GameView(Context context) {
        super(context);
        init();
    }

    public GameView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public GameView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    // ── Callback interfaces ────────────────────────────────────────────────────

    public interface SwapListener {
        void onSwapAttempted(boolean valid);
    }

    // ── State ──────────────────────────────────────────────────────────────────

    private GameEngine engine;
    private SwapListener swapListener;

    // Selection
    private int selRow = -1;
    private int selCol = -1;

    // Hint
    private int[] hintCells = null;       // {r1,c1,r2,c2}
    private boolean hintVisible = false;
    private final Handler hintHandler = new Handler(Looper.getMainLooper());
    private final Runnable hideHint = () -> {
        hintCells = null;
        hintVisible = false;
        invalidate();
    };

    // Layout
    private float cellSize;
    private float boardLeft;
    private float boardTop;

    // Paints
    private Paint cellPaint;
    private Paint selectedPaint;
    private Paint hintPaint;
    private Paint lockedPaint;
    private Paint textPaint;
    private Paint overlayPaint;

    // ── Init ───────────────────────────────────────────────────────────────────

    private void init() {
        cellPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
        cellPaint.setStyle(Paint.Style.FILL);

        selectedPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
        selectedPaint.setStyle(Paint.Style.STROKE);
        selectedPaint.setColor(Color.WHITE);
        selectedPaint.setStrokeWidth(5f);

        hintPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
        hintPaint.setStyle(Paint.Style.STROKE);
        hintPaint.setColor(0xFFFFD700);
        hintPaint.setStrokeWidth(5f);

        lockedPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
        lockedPaint.setStyle(Paint.Style.FILL);
        lockedPaint.setColor(0xFF546E7A);

        textPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
        textPaint.setTextAlign(Paint.Align.CENTER);

        overlayPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
        overlayPaint.setStyle(Paint.Style.FILL);
    }

    // ── Public API ─────────────────────────────────────────────────────────────

    public void setEngine(GameEngine engine) {
        this.engine = engine;
        invalidate();
    }

    public void setSwapListener(SwapListener listener) {
        this.swapListener = listener;
    }

    public void showHint() {
        if (engine == null) return;
        hintCells = engine.findHint();
        hintVisible = hintCells != null;
        hintHandler.removeCallbacks(hideHint);
        if (hintVisible) hintHandler.postDelayed(hideHint, 2000);
        invalidate();
    }

    public void cancelHint() {
        hintHandler.removeCallbacks(hideHint);
        hintCells = null;
        hintVisible = false;
    }

    // ── Layout ─────────────────────────────────────────────────────────────────

    @Override
    protected void onSizeChanged(int w, int h, int oldw, int oldh) {
        super.onSizeChanged(w, h, oldw, oldh);
        if (engine == null) return;
        recalcLayout(w, h);
    }

    private void recalcLayout(int w, int h) {
        int size = engine.getBoardSize();
        float available = Math.min(w, h);
        cellSize = available / size;
        boardLeft = (w - cellSize * size) / 2f;
        boardTop = (h - cellSize * size) / 2f;
    }

    // ── Draw ───────────────────────────────────────────────────────────────────

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        if (engine == null) return;

        int size = engine.getBoardSize();
        float pad = cellSize * 0.06f;
        float radius = cellSize * 0.18f;
        float emojiSize = cellSize * 0.56f;
        textPaint.setTextSize(emojiSize);

        RectF rect = new RectF();

        for (int r = 0; r < size; r++) {
            for (int c = 0; c < size; c++) {
                Cell cell = engine.getCell(r, c);
                float left = boardLeft + c * cellSize + pad;
                float top = boardTop + r * cellSize + pad;
                float right = left + cellSize - 2 * pad;
                float bottom = top + cellSize - 2 * pad;
                rect.set(left, top, right, bottom);

                if (cell.isLocked) {
                    drawLockedCell(canvas, rect, radius, emojiSize);
                } else if (cell.isEmpty()) {
                    overlayPaint.setColor(0x22FFFFFF);
                    canvas.drawRoundRect(rect, radius, radius, overlayPaint);
                } else {
                    drawFruitCell(canvas, rect, radius, emojiSize, cell.fruitIndex);
                }

                // Selected highlight
                if (r == selRow && c == selCol) {
                    canvas.drawRoundRect(rect, radius, radius, selectedPaint);
                }

                // Hint highlight
                if (hintVisible && hintCells != null) {
                    if ((r == hintCells[0] && c == hintCells[1])
                            || (r == hintCells[2] && c == hintCells[3])) {
                        canvas.drawRoundRect(rect, radius, radius, hintPaint);
                    }
                }
            }
        }
    }

    private void drawFruitCell(Canvas canvas, RectF rect, float radius, float emojiSize, int fruitIndex) {
        int baseColor = GameEngine.FRUIT_COLORS[Math.min(fruitIndex, GameEngine.FRUIT_COLORS.length - 1)];
        // Semi-transparent background tinted to fruit colour
        cellPaint.setColor(Color.argb(70,
                Color.red(baseColor), Color.green(baseColor), Color.blue(baseColor)));
        canvas.drawRoundRect(rect, radius, radius, cellPaint);

        // Emoji
        String emoji = GameEngine.FRUIT_EMOJIS[Math.min(fruitIndex, GameEngine.FRUIT_EMOJIS.length - 1)];
        float cx = rect.centerX();
        float cy = rect.centerY() + emojiSize * 0.35f;
        canvas.drawText(emoji, cx, cy, textPaint);
    }

    private void drawLockedCell(Canvas canvas, RectF rect, float radius, float emojiSize) {
        canvas.drawRoundRect(rect, radius, radius, lockedPaint);
        textPaint.setTextSize(emojiSize * 0.75f);
        canvas.drawText("🧊", rect.centerX(), rect.centerY() + emojiSize * 0.3f, textPaint);
        textPaint.setTextSize(emojiSize); // restore
    }

    // ── Touch ──────────────────────────────────────────────────────────────────

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        if (engine == null) return true;
        if (engine.getState().isGameOver || engine.getState().isPaused) return true;
        if (event.getAction() != MotionEvent.ACTION_DOWN) return true;

        int col = (int) ((event.getX() - boardLeft) / cellSize);
        int row = (int) ((event.getY() - boardTop) / cellSize);
        int sz = engine.getBoardSize();

        if (row < 0 || row >= sz || col < 0 || col >= sz) {
            clearSelection();
            return true;
        }

        Cell cell = engine.getCell(row, col);
        if (cell.isLocked || cell.isEmpty()) {
            clearSelection();
            return true;
        }

        if (selRow == -1) {
            // First tap: select
            selRow = row;
            selCol = col;
            cancelHint();
            invalidate();
        } else if (selRow == row && selCol == col) {
            // Same cell: deselect
            clearSelection();
        } else {
            int dr = Math.abs(row - selRow);
            int dc = Math.abs(col - selCol);
            if ((dr == 1 && dc == 0) || (dr == 0 && dc == 1)) {
                // Adjacent: attempt swap
                boolean valid = engine.trySwap(selRow, selCol, row, col);
                if (swapListener != null) swapListener.onSwapAttempted(valid);
                clearSelection();
                invalidate();
            } else {
                // Non-adjacent: re-select new cell
                selRow = row;
                selCol = col;
                invalidate();
            }
        }
        return true;
    }

    private void clearSelection() {
        selRow = -1;
        selCol = -1;
        invalidate();
    }
}

File: app/src/main/java/com/fruitblast/game/ui/GameView.kt
package com.fruitblast.game.ui

import android.animation.ValueAnimator
import android.content.Context
import android.graphics.*
import android.os.Handler
import android.os.Looper
import android.view.MotionEvent
import android.view.View
import android.view.animation.BounceInterpolator
import android.view.animation.DecelerateInterpolator
import com.fruitblast.game.data.*
import com.fruitblast.game.engine.GameEngine
import com.fruitblast.game.engine.GameEventListener
import kotlin.math.abs
import kotlin.math.min

/**
 * Custom View that renders the match-3 game board using Canvas APIs.
 *
 * Responsibilities:
 * - Drawing fruit cells, grid, highlights, special indicators
 * - Handling touch input (tap-to-select + tap-to-swap)
 * - Running animations: match flash, gravity fall, spawn pop, swap wobble
 * - Showing hint highlights
 */
class GameView(context: Context) : View(context), GameEventListener {

    // ─── Engine ───────────────────────────────────────────────────────────────
    lateinit var engine: GameEngine
        private set

    fun attachEngine(eng: GameEngine) {
        engine = eng
        eng.listener = this
    }

    // ─── Callbacks for Activity ───────────────────────────────────────────────
    var onScoreChanged: ((Int) -> Unit)? = null
    var onComboChanged: ((Int) -> Unit)? = null
    var onGameOver: ((GameState) -> Unit)? = null
    var onSwapSound: (() -> Unit)? = null
    var onMatchSound: (() -> Unit)? = null
    var onInvalidSound: (() -> Unit)? = null
    var onShuffleSound: (() -> Unit)? = null

    // ─── Selection State ──────────────────────────────────────────────────────
    private var selectedRow = -1
    private var selectedCol = -1
    private var isAnimating = false
    private var hintRow1 = -1; private var hintCol1 = -1
    private var hintRow2 = -1; private var hintCol2 = -1

    // ─── Layout ───────────────────────────────────────────────────────────────
    private var cellSize = 0f
    private var boardOffsetX = 0f
    private var boardOffsetY = 0f
    private val cellPadding get() = cellSize * 0.07f

    // ─── Paints ───────────────────────────────────────────────────────────────
    private val cellBgPaint = Paint(Paint.ANTI_ALIAS_FLAG)
    private val selectedPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        style = Paint.Style.STROKE
        strokeWidth = 6f
        color = Color.WHITE
    }
    private val hintPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        style = Paint.Style.STROKE
        strokeWidth = 5f
        color = Color.YELLOW
    }
    private val emojiPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        textAlign = Paint.Align.CENTER
    }
    private val lockedPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        color = Color.parseColor("#A0B8D8E0")
    }
    private val shadowPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        color = Color.parseColor("#33000000")
    }
    private val specialBadgePaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        textAlign = Paint.Align.CENTER
    }
    private val glowPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        style = Paint.Style.FILL
        maskFilter = BlurMaskFilter(20f, BlurMaskFilter.Blur.SOLID)
    }

    // Per-cell animation state (alpha for flash, scaleY for fall)
    private val cellAnimAlpha = Array(9) { FloatArray(9) { 1f } }
    private val cellAnimScale = Array(9) { FloatArray(9) { 1f } }
    private val cellAnimOffsetY = Array(9) { FloatArray(9) { 0f } }

    // Hint pulse animation
    private var hintPulseAlpha = 1f
    private val hintAnimator = ValueAnimator.ofFloat(0.3f, 1f).apply {
        duration = 600
        repeatMode = ValueAnimator.REVERSE
        repeatCount = ValueAnimator.INFINITE
        addUpdateListener { hintPulseAlpha = it.animatedValue as Float; invalidate() }
    }

    private val handler = Handler(Looper.getMainLooper())

    // ─── Layout ───────────────────────────────────────────────────────────────

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        super.onSizeChanged(w, h, oldw, oldh)
        if (!::engine.isInitialized) return
        computeLayout(w, h)
    }

    private fun computeLayout(w: Int, h: Int) {
        val rows = engine.rows
        val cols = engine.cols
        val maxCellW = w.toFloat() / cols
        val maxCellH = h.toFloat() / rows
        cellSize = min(maxCellW, maxCellH) * 0.97f
        boardOffsetX = (w - cellSize * cols) / 2f
        boardOffsetY = (h - cellSize * rows) / 2f
        emojiPaint.textSize = cellSize * 0.58f
        specialBadgePaint.textSize = cellSize * 0.25f
    }

    // ─── Drawing ──────────────────────────────────────────────────────────────

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        if (!::engine.isInitialized) return

        drawBackground(canvas)
        drawCells(canvas)
        drawSelectionHighlight(canvas)
        drawHintHighlight(canvas)
    }

    private fun drawBackground(canvas: Canvas) {
        val bgPaint = Paint().apply { color = Color.parseColor("#1A1A2E") }
        canvas.drawRect(0f, 0f, width.toFloat(), height.toFloat(), bgPaint)
        // Subtle grid lines
        val gridPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
            color = Color.parseColor("#22FFFFFF")
            strokeWidth = 1f
        }
        val rows = engine.rows; val cols = engine.cols
        for (r in 0..rows) {
            val y = boardOffsetY + r * cellSize
            canvas.drawLine(boardOffsetX, y, boardOffsetX + cols * cellSize, y, gridPaint)
        }
        for (c in 0..cols) {
            val x = boardOffsetX + c * cellSize
            canvas.drawLine(x, boardOffsetY, x, boardOffsetY + rows * cellSize, gridPaint)
        }
    }

    private fun drawCells(canvas: Canvas) {
        val rows = engine.rows; val cols = engine.cols
        for (r in 0 until rows) {
            for (c in 0 until cols) {
                drawCell(canvas, r, c)
            }
        }
    }

    private fun drawCell(canvas: Canvas, row: Int, col: Int) {
        val cell = engine.board[row][col]
        val left = boardOffsetX + col * cellSize + cellPadding
        val top = boardOffsetY + row * cellSize + cellPadding + cellAnimOffsetY[row][col]
        val right = left + cellSize - cellPadding * 2
        val bottom = top + cellSize - cellPadding * 2

        val cx = (left + right) / 2f
        val cy = (top + bottom) / 2f
        val radius = (cellSize / 2f) - cellPadding

        val alpha = (cellAnimAlpha[row][col] * 255).toInt().coerceIn(0, 255)
        val scale = cellAnimScale[row][col]

        canvas.save()
        canvas.scale(scale, scale, cx, cy)

        // Locked ice cell
        if (cell.isLocked) {
            lockedPaint.alpha = alpha
            canvas.drawRoundRect(left, top, right, bottom, radius * 0.3f, radius * 0.3f, lockedPaint)
            emojiPaint.alpha = alpha
            canvas.drawText("🧊", cx, cy + emojiPaint.textSize * 0.35f, emojiPaint)
            canvas.restore()
            return
        }

        val fruitType = cell.fruitType ?: run { canvas.restore(); return }

        // Drop shadow
        shadowPaint.alpha = (alpha * 0.4f).toInt()
        canvas.drawCircle(cx + 3f, cy + 5f, radius, shadowPaint)

        // Glow for special fruits
        if (cell.isSpecial) {
            glowPaint.color = Color.argb((alpha * 0.6f).toInt(), 255, 255, 100)
            canvas.drawCircle(cx, cy, radius + 8f, glowPaint)
        }

        // Background circle
        cellBgPaint.color = fruitType.bgColor
        cellBgPaint.alpha = alpha
        canvas.drawCircle(cx, cy, radius, cellBgPaint)

        // Highlight shine (top-left)
        val shinePaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
            color = Color.WHITE
            this.alpha = (alpha * 0.3f).toInt()
        }
        canvas.drawCircle(cx - radius * 0.25f, cy - radius * 0.25f, radius * 0.35f, shinePaint)

        // Emoji fruit
        emojiPaint.alpha = alpha
        val textY = cy + emojiPaint.textSize * 0.35f
        canvas.drawText(fruitType.emoji, cx, textY, emojiPaint)

        // Special badge overlay
        if (cell.isSpecial) {
            val badge = when (cell.specialType) {
                SpecialType.ROW_BOMB      -> "↔"
                SpecialType.COL_BOMB      -> "↕"
                SpecialType.CROSS_BOMB    -> "✚"
                SpecialType.RAINBOW_FRUIT -> "★"
                else -> ""
            }
            if (badge.isNotEmpty()) {
                val badgePaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
                    color = Color.WHITE
                    textSize = cellSize * 0.28f
                    textAlign = Paint.Align.CENTER
                    typeface = Typeface.DEFAULT_BOLD
                    this.alpha = alpha
                }
                canvas.drawText(badge, cx, bottom - cellPadding * 1.5f, badgePaint)
            }
        }

        canvas.restore()
    }

    private fun drawSelectionHighlight(canvas: Canvas) {
        if (selectedRow < 0 || selectedCol < 0) return
        val cx = boardOffsetX + selectedCol * cellSize + cellSize / 2f
        val cy = boardOffsetY + selectedRow * cellSize + cellSize / 2f
        val radius = cellSize / 2f - cellPadding

        val ringPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
            style = Paint.Style.STROKE
            strokeWidth = 5f
            color = Color.WHITE
            alpha = 220
        }
        canvas.drawCircle(cx, cy, radius + 4f, ringPaint)

        val glowRingPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
            style = Paint.Style.STROKE
            strokeWidth = 10f
            color = Color.parseColor("#80FFFFFF")
        }
        canvas.drawCircle(cx, cy, radius + 8f, glowRingPaint)
    }

    private fun drawHintHighlight(canvas: Canvas) {
        if (hintRow1 < 0) return
        drawHintRing(canvas, hintRow1, hintCol1)
        drawHintRing(canvas, hintRow2, hintCol2)
    }

    private fun drawHintRing(canvas: Canvas, row: Int, col: Int) {
        val cx = boardOffsetX + col * cellSize + cellSize / 2f
        val cy = boardOffsetY + row * cellSize + cellSize / 2f
        val radius = cellSize / 2f - cellPadding
        hintPaint.alpha = (hintPulseAlpha * 255).toInt()
        hintPaint.color = Color.YELLOW
        canvas.drawCircle(cx, cy, radius + 4f, hintPaint)
    }

    // ─── Touch Handling ───────────────────────────────────────────────────────

    override fun onTouchEvent(event: MotionEvent): Boolean {
        if (event.action != MotionEvent.ACTION_UP) return true
        if (isAnimating) return true
        if (!::engine.isInitialized) return true
        if (engine.gameState.isGameOver || engine.gameState.isPaused) return true

        val touchX = event.x
        val touchY = event.y
        val col = ((touchX - boardOffsetX) / cellSize).toInt()
        val row = ((touchY - boardOffsetY) / cellSize).toInt()

        if (row < 0 || row >= engine.rows || col < 0 || col >= engine.cols) {
            clearSelection()
            return true
        }

        clearHint()

        if (selectedRow < 0) {
            // First tap: select
            if (!engine.board[row][col].isEmpty && !engine.board[row][col].isLocked) {
                selectedRow = row; selectedCol = col
                animateCellPop(row, col)
                invalidate()
            }
        } else {
            // Second tap: attempt swap or change selection
            if (row == selectedRow && col == selectedCol) {
                clearSelection()
            } else if (isAdjacent(selectedRow, selectedCol, row, col)) {
                val r1 = selectedRow; val c1 = selectedCol
                clearSelection()
                animateSwapAttempt(r1, c1, row, col)
            } else {
                // Re-select
                if (!engine.board[row][col].isEmpty && !engine.board[row][col].isLocked) {
                    selectedRow = row; selectedCol = col
                    animateCellPop(row, col)
                    invalidate()
                } else {
                    clearSelection()
                }
            }
        }
        return true
    }

    private fun isAdjacent(r1: Int, c1: Int, r2: Int, c2: Int): Boolean {
        val dr = abs(r1 - r2); val dc = abs(c1 - c2)
        return (dr == 1 && dc == 0) || (dr == 0 && dc == 1)
    }

    private fun clearSelection() {
        selectedRow = -1; selectedCol = -1
        invalidate()
    }

    // ─── Animations ───────────────────────────────────────────────────────────

    private fun animateCellPop(row: Int, col: Int) {
        ValueAnimator.ofFloat(1f, 1.2f, 1f).apply {
            duration = 200
            addUpdateListener {
                cellAnimScale[row][col] = it.animatedValue as Float
                invalidate()
            }
            start()
        }
    }

    private fun animateSwapAttempt(r1: Int, c1: Int, r2: Int, c2: Int) {
        isAnimating = true
        onSwapSound?.invoke()

        // Visual swap tween
        val dx = (c2 - c1) * cellSize
        val dy = (r2 - r1) * cellSize

        val swapAnim = ValueAnimator.ofFloat(0f, 1f).apply {
            duration = 180
            interpolator = DecelerateInterpolator()
            addUpdateListener { anim ->
                val frac = anim.animatedValue as Float
                cellAnimOffsetY[r1][c1] = dy * frac
                cellAnimOffsetY[r2][c2] = -dy * frac
                invalidate()
            }
        }
        swapAnim.start()
        handler.postDelayed({
            cellAnimOffsetY[r1][c1] = 0f
            cellAnimOffsetY[r2][c2] = 0f

            val success = engine.trySwap(r1, c1, r2, c2)
            if (!success) {
                // Snap back (engine already reverted swap)
                animateSnapBack(r1, c1, r2, c2)
            } else {
                isAnimating = false
                invalidate()
                notifyUIUpdate()
            }
        }, 200)
    }

    private fun animateSnapBack(r1: Int, c1: Int, r2: Int, c2: Int) {
        onInvalidSound?.invoke()
        ValueAnimator.ofFloat(0f, 0.15f, -0.1f, 0f).apply {
            duration = 280
            addUpdateListener { anim ->
                val frac = anim.animatedValue as Float
                cellAnimScale[r1][c1] = 1f + frac
                cellAnimScale[r2][c2] = 1f + frac
                invalidate()
            }
            start()
        }
        handler.postDelayed({
            cellAnimScale[r1][c1] = 1f
            cellAnimScale[r2][c2] = 1f
            isAnimating = false
            invalidate()
        }, 300)
    }

    private fun animateMatchFlash(cells: List<Pair<Int, Int>>) {
        onMatchSound?.invoke()
        ValueAnimator.ofFloat(1f, 0.2f, 1f, 0f).apply {
            duration = 350
            addUpdateListener { anim ->
                val alpha = anim.animatedValue as Float
                for ((r, c) in cells) cellAnimAlpha[r][c] = alpha
                invalidate()
            }
            start()
        }
    }

    private fun animateGravityFall() {
        ValueAnimator.ofFloat(-cellSize * 0.5f, 0f).apply {
            duration = 280
            interpolator = BounceInterpolator()
            addUpdateListener { anim ->
                val offset = anim.animatedValue as Float
                for (r in 0 until engine.rows) {
                    for (c in 0 until engine.cols) {
                        if (!engine.board[r][c].isEmpty) {
                            cellAnimOffsetY[r][c] = offset
                        }
                    }
                }
                invalidate()
            }
            start()
        }
    }

    // ─── Hint System ──────────────────────────────────────────────────────────

    fun showHint() {
        val hint = engine.getHint() ?: return
        hintRow1 = hint.first.first;  hintCol1 = hint.first.second
        hintRow2 = hint.second.first; hintCol2 = hint.second.second
        if (!hintAnimator.isRunning) hintAnimator.start()
        invalidate()

        // Auto-clear hint after 3 seconds
        handler.postDelayed({ clearHint() }, 3000)
    }

    private fun clearHint() {
        hintAnimator.cancel()
        hintRow1 = -1; hintCol1 = -1; hintRow2 = -1; hintCol2 = -1
        invalidate()
    }

    // ─── GameEventListener ────────────────────────────────────────────────────

    override fun onValidSwap(r1: Int, c1: Int, r2: Int, c2: Int) { /* handled in touch */ }
    override fun onInvalidSwap(r1: Int, c1: Int, r2: Int, c2: Int) { /* handled in touch */ }

    override fun onCellsMatched(cells: List<Pair<Int, Int>>, combo: Int, score: Int) {
        animateMatchFlash(cells)
        // Reset matched cells alpha after animation
        handler.postDelayed({
            for ((r, c) in cells) {
                cellAnimAlpha[r][c] = 1f
                cellAnimScale[r][c] = 1f
                cellAnimOffsetY[r][c] = 0f
            }
        }, 360)
        notifyUIUpdate()
    }

    override fun onCascadeStart() {
        handler.postDelayed({ animateGravityFall() }, 100)
    }

    override fun onCascadeEnd() {
        resetAllAnimStates()
        isAnimating = false
        invalidate()
        notifyUIUpdate()
    }

    override fun onNoMovesAvailable() {
        onShuffleSound?.invoke()
    }

    override fun onBoardShuffled() {
        resetAllAnimStates()
        invalidate()
    }

    override fun onGameOver(state: GameState) {
        handler.postDelayed({
            onGameOver?.invoke(state)
        }, 500)
    }

    private fun resetAllAnimStates() {
        for (r in 0 until engine.rows) {
            for (c in 0 until engine.cols) {
                cellAnimAlpha[r][c] = 1f
                cellAnimScale[r][c] = 1f
                cellAnimOffsetY[r][c] = 0f
            }
        }
    }

    private fun notifyUIUpdate() {
        if (::engine.isInitialized) {
            onScoreChanged?.invoke(engine.gameState.score)
            onComboChanged?.invoke(engine.gameState.combo)
        }
    }

    fun cleanup() {
        hintAnimator.cancel()
        handler.removeCallbacksAndMessages(null)
    }
}
